src/lineup/internal/cmds/filter.ts

```diff
@@ -0,0 +1,126 @@
+ * @param filter Any value that could be a filter
+ */
+export function isSerializedFilter(filter: any): ISerializableLineUpFilter {
+  return filter && filter.hasOwnProperty('filter');
+}
+
+/**
+ * Serializes LineUp string filter, which can contain RegExp objects to an IRegexFilter object.
+ * The return value of this function can be passed to `JSON.stringify()` and stored in the provenance graph.
+ *
+ * Background information:
+ * The serialization step is necessary, because RegExp objects are converted into an empty object `{}` on `JSON.stringify`.
+ * ```
+ * JSON.stringify(/^123$/gm); // result: {}
+ * ```
+ *
+ * @internal
+ *
+ * @param filter LineUp filter object
+ * @returns Returns the `ISerializableLineUpFilter` object
+ */
+export function serializeLineUpFilter(filter: ILineUpStringFilter): ISerializableLineUpFilter {
+  const value = filter.filter;
+  const isRegexp = value instanceof RegExp;
+  return {
+    filter: {
+      value: isRegexp ? value.toString() : value as ILineUpStringFilterValue,
+      isRegExp: isRegexp
+    },
+    filterMissing: filter.filterMissing
+  };
+}
+
+/**
+ * Restores a RegExp object from a given IRegExpFilter object.
+ * In case a string is passed to this function no deserialization is applied.
+ *
+ * @interal
+ * @param filter Filter as string or plain object matching the IRegExpFilter
+ * @param filterMissing The flag indicates if missing values should be filtered (default = `false`)
+ * @returns Returns the input string or the restored RegExp object
+ */
+export function restoreLineUpFilter(filter: ILineUpStringFilterValue | IRegExpFilter | ISerializableLineUpFilter, filterMissing = false): ILineUpStringFilter {
+  const isSimpleFilter = (filter: ILineUpStringFilterValue | IRegExpFilter | ISerializableLineUpFilter): filter is ILineUpStringFilterValue => filter === null || typeof filter === 'string' || Array.isArray(filter);
+  const isIRegExpFilter = ((filter: IRegExpFilter | ISerializableLineUpFilter): filter is IRegExpFilter => filter.hasOwnProperty('isRegExp'));
+  const isISerializableLineUpFilter = (filter: IRegExpFilter | ISerializableLineUpFilter): filter is ISerializableLineUpFilter => filter.hasOwnProperty('filterMissing');
+
+  if (isSimpleFilter(filter)) {
+    return {filter, filterMissing};
+
+  } else if (isIRegExpFilter(filter)) {
+    if (filter.isRegExp) {
+      const serializedRegexParser = /^\/(.+)\/(\w+)?$/; // from https://gist.github.com/tenbits/ec7f0155b57b2d61a6cc90ef3d5f8b49
+      const matches = serializedRegexParser.exec(filter.value as string);
+      const [_full, regexString, regexFlags] = matches;
+      return {filter: new RegExp(regexString, regexFlags), filterMissing};
+    }
+
+    return restoreLineUpFilter(filter.value, filterMissing);
+
+  } else if (isISerializableLineUpFilter(filter)) {
+    return restoreLineUpFilter(filter.filter, filter.filterMissing);
+
+  }
+
+  throw new Error('Unknown LineUp filter format. Unable to restore the given filter.');
```

Comment:
I added this error message to prevent cases where the return value of this function is `void`. After playing around with this implementation I got an error with the regexp string:

![ordino-lineup-unknown-filter-format](https://user-images.githubusercontent.com/5851088/78226481-0e486f80-74cc-11ea-835f-c709792ad226.gif)

This is the error in the console:

```
filter.ts:125 Uncaught (in promise) Error: Unknown LineUp filter format. Unable to restore the given filter.
    at restoreLineUpFilter (filter.ts:125)
    at restoreLineUpFilter (filter.ts:121)
    at ActionNode.<anonymous> (cmds.ts:227)
    at Generator.next (<anonymous>)
    at fulfilled (tslib.es6.js:71)
```

@oltionchampari Can you check why calling it recursively (starting from line 121) and not returning `void` (without the error) and now we run into the error? In guess is that we still have an uncovered case where the three typeguards above are not working correctly.

My expectation would be that we do not get the error message and one of the typeguards catches the recursive call and returns a valid filter object.

Reply:
The issue was that when you go back in the provenance graph the `restoreLineUpFilter()`
function gets called with the old value which gets derived from the Ranking.Thus it hasn't been serialized:

 ```typescript
const fromGraphFilter = {filter: /abc/, filterMissing: false};
   restoreLineUpFilter(fromGraphFilter)
```
I added a test to cover this case and returned the filter when it's already serialized.